servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,342 @@
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeCleanup(channel);
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> decode(final Publisher<Buffer> from, final BufferAllocator allocator) {
+        requireNonNull(from);
+        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+            private final ByteToMessageDecoder decoder = decoderSupplier.get();
+            private final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+            @Nullable
+            PublisherSource.Subscription subscription;
+
+            @Override
+            public void onSubscribe(final PublisherSource.Subscription subscription) {
+                this.subscription = subscription;
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert subscription != null;
+                if (!channel.isOpen()) {
+                    throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
+                }
+
+                if (src == null) {
+                    throw new NullPointerException();
+                }
+
+                // onNext will produce AT-MOST N items (as received)
+                channel.writeInbound(toByteBuf(src));
+                Buffer buffer = drainChannelQueueToSingleBuffer(channel.inboundMessages(), allocator);
+
+                if (buffer != null && buffer.readableBytes() > 0) {
+                    subscriber.onNext(buffer);
+                } else {
+                    // Not enough data to decompress, ask for more
+                    subscription.request(1);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                safeCleanup(channel);
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                try {
+                    cleanup(channel);
+                } catch (Throwable t) {
+                    subscriber.onError(t);
+                    return;
+                }
+
+                subscriber.onComplete();
+            }
+        });
+    }
+
+    private EmbeddedChannel newEmbeddedChannel(final ChannelHandler handler, final BufferAllocator allocator) {
+        final EmbeddedChannel channel = new EmbeddedChannel(handler);
+        channel.config().setAllocator(getByteBufAllocator(allocator));
+        return channel;
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer) {
+        return BufferUtils.extractByteBufOrCreate(buffer);
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer, final int offset, final int length) {
+        return BufferUtils.extractByteBufOrCreate(buffer.readerIndex(buffer.readerIndex() + offset).readSlice(length));
```

Comment:
We agreed to keep a uniform API across the board with offsets too, I think we should stick to it. The slice here is only an exceptional case and not the norm, most of the times we expect the buffers to be Netty buffers.

Reply:
Even if it's a netty buffer, we need a slice because we can not silently skip data between `readerIndex` and `readerIndex + offset`:
```java
Bufffer b = fromAscii("Hi Thomas");
Buffer compressed = codec.encode(b, 3, 6);
```
The "Hi " part is lost now. I don't think any other API that takes an offset behaves in the same way.